### PR TITLE
[PyTorch] Further reduce cost of TypeMeta::_typeMetaData (by 10x!)

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98105

Currently we should be paying a small cost for the
thread-safe initialization of `index`. Now we should eliminate that
cost. (10x figure in the title comes from internal benchmark that just
calls `TypeMeta::Match<caffe2::Tensor>()` in a loop).

Differential Revision: [D44597852](https://our.internmc.facebook.com/intern/diff/D44597852/)

### CODE REVIEW

**Comment by swolchok** [comment]
previously: #83707

File: c10/util/typeid.h [hunk 1]
```diff
@@ -520,6 +520,7 @@ class C10_API TypeMeta final {
 
   static uint16_t existingMetaDataIndexForType(TypeIdentifier identifier);
 
+ public:
```
Comment (swolchok) [inline c10/util/typeid.h:523]: I debated adding _DO_NOT_CALL_DIRECTLY to the name of addTypeMetaData, but that didn't seem very PyTorchy.

Comment (ezyang) [reply c10/util/typeid.h:523]: meh, I don't mind haha

**Review by ezyang — APPROVED** [review]

**Comment by swolchok** [comment]
legit windows failures that I missed -- `error LNK2019: unresolved external symbol "unsigned short const caffe2::detail::std_string_metadata_index"` and similar

**Comment by weiwangmeta** [comment]
[I manually cancelled via this job: ](https://github.com/pytorch/pytorch/actions/runs/4641203578)

**Comment by swolchok** [comment]
macos-12-py3-arm64 failure looks unrelated. will try having it rebase before merge

**Pull request closed without merging**